File: 02.py
with open('02-input', 'r') as f:
    data = f.readlines()

def parse(d):
    d = d.split(':')
    pw = d[1][1:-1]
    policy = d[0]
    letter = policy[-1]
    min_max_val = policy.split()[0].split('-')
    min_val = int(min_max_val[0])
    max_val = int(min_max_val[1])
    return min_val, max_val, letter, pw


def verifyA(min_val, max_val, letter, pw):
    count = pw.count(letter)
    if min_val <= count and count <= max_val:
        return 1
    else:
        return 0

def verifyB(min_val, max_val, letter, pw):
    letter1 = pw[min_val - 1]
    letter2 = pw[max_val - 1]
    if letter1 == letter2:
        return 0
    elif letter1 == letter or letter2 == letter:
        return 1
    else:
        return 0
    
validA = 0
validB = 0

for d in data:
    min_val, max_val, letter, pw = parse(d)
    validA += verifyA(min_val, max_val, letter, pw)
    validB += verifyB(min_val, max_val, letter, pw)

print("Solution for A:", validA)
print("Solution for A:", validB)

File: 02.go
package main

import (
    "fmt"
    "io/ioutil"
    "strconv"
    "strings"
)

type PasswordPolicy struct {
    minVal int
    maxVal int
    letter string
    pw string
}

func readInput() []PasswordPolicy {
    data, err := ioutil.ReadFile("02-input")
    if err != nil {
        panic(err)
    }

    lines := strings.Split(string(data), "\n")
    lines = lines[:len(lines) - 1]
    pws := make([]PasswordPolicy, 0, len(lines))

    for _, line := range lines {
        lineSplit := strings.Split(line, ":")
        policy := lineSplit[0]
        pw := lineSplit[1]
        pw = pw[1:]
        letter := string(policy[len(policy) - 1])
        minMaxVal := strings.Split(strings.Split(policy, " ")[0], "-") 
        minVal, _ := strconv.Atoi(minMaxVal[0])
        maxVal, _ := strconv.Atoi(minMaxVal[1])
        pwPolicy := PasswordPolicy{minVal, maxVal, letter, pw}
        pws = append(pws, pwPolicy)
    }

    return pws
}

func partA(pws []PasswordPolicy) int {
    
    result := 0

    for _, pwPolicy := range pws {
        count := strings.Count(pwPolicy.pw, pwPolicy.letter)
        if pwPolicy.minVal <= count && count <= pwPolicy.maxVal {
            result += 1
        }
    }

    return result
}

func partB(pws []PasswordPolicy) int {

    result := 0

    for _, pwPolicy := range pws {
        letter1 := string(pwPolicy.pw[pwPolicy.minVal - 1])
        letter2 := string(pwPolicy.pw[pwPolicy.maxVal - 1])
        if letter1 == pwPolicy.letter && letter2 != pwPolicy.letter {
            result += 1
        } else if letter1 != pwPolicy.letter && letter2 == pwPolicy.letter {
            result += 1
        }
    }

    return result
}

func main() {
    pws := readInput()
    ansA := partA(pws)
    fmt.Printf("Solution for A: %d\n", ansA)
    ansB := partB(pws)
    fmt.Printf("Solution for B: %d\n", ansB)

}

